tracer/src/Datadog.Trace/LifetimeManager.cs

```diff
@@ -109,17 +95,55 @@ private void CurrentDomain_DomainUnload(object sender, EventArgs e)
 
         private void RunShutdownTasks()
         {
+            var current = SynchronizationContext.Current;
             try
             {
-                while (_shutdownHooks.TryDequeue(out var action))
+                if (current is not null)
+                {
+                    SetSynchronizationContext(null);
+                }
+
+                while (_shutdownHooks.TryDequeue(out var actionOrFunc))
                 {
-                    action();
+                    if (actionOrFunc is Action action)
+                    {
+                        action();
+                    }
+                    else if (actionOrFunc is Func<Task> func)
+                    {
+                        func().Wait(TaskTimeout);
+                    }
```

Comment:
```suggestion
                    } else {
                        Log.Error("Something terrible happened");
                    }
```
Since `_shutdownHooks` can contain any kind of `object`, this could happen by mistake.

Reply:
for that the public api must change, checked added on https://github.com/DataDog/dd-trace-dotnet/pull/3075/commits/2d31032e4cd133dc0f681848da4ffaa1fc722b7b